test/e2e/specs/editor/various/block-editor-keyboard-shortcuts.spec.js

```diff
@@ -0,0 +1,220 @@
+					attributes: { content: '2nd' },
+				},
+			] );
+		} );
+
+		test( 'should move the block down', async ( {
+			editor,
+			page,
+			pageUtils,
+		} ) => {
+			await addTestParagraphBlocks( { editor, page } );
+			await page.keyboard.press( 'ArrowUp' );
+			await pageUtils.pressKeys( 'secondary+y' );
+
+			await expect.poll( editor.getBlocks ).toMatchObject( [
+				{
+					name: 'core/paragraph',
+					attributes: { content: '1st' },
+				},
+				{
+					name: 'core/paragraph',
+					attributes: { content: '3rd' },
+				},
+				{
+					name: 'core/paragraph',
+					attributes: { content: '2nd' },
+				},
+			] );
+		} );
+	} );
+
+	test.describe( 'move blocks - multiple blocks selected', () => {
+		test( 'should move the blocks up', async ( {
+			editor,
+			page,
+			pageUtils,
+		} ) => {
+			await addTestParagraphBlocks( { editor, page } );
+			await pageUtils.pressKeys( 'shift+ArrowUp' );
+			await expect(
+				page
+					.getByRole( 'toolbar', { name: 'Block tools' } )
+					.getByRole( 'button', { name: 'Multiple blocks selected' } )
+			).toBeVisible();
+			await pageUtils.pressKeys( 'secondary+t' );
+
+			await expect.poll( editor.getBlocks ).toMatchObject( [
+				{
+					name: 'core/paragraph',
+					attributes: { content: '2nd' },
+				},
+				{
+					name: 'core/paragraph',
+					attributes: { content: '3rd' },
+				},
+				{
+					name: 'core/paragraph',
+					attributes: { content: '1st' },
+				},
+			] );
+		} );
+
+		test( 'should move the blocks down', async ( {
+			editor,
+			page,
+			pageUtils,
+		} ) => {
+			await addTestParagraphBlocks( { editor, page } );
+			await page.keyboard.press( 'ArrowUp' );
+			await pageUtils.pressKeys( 'shift+ArrowUp' );
+			await expect(
+				page
+					.getByRole( 'toolbar', { name: 'Block tools' } )
+					.getByRole( 'button', { name: 'Multiple blocks selected' } )
+			).toBeVisible();
+			await pageUtils.pressKeys( 'secondary+y' );
+
+			await expect.poll( editor.getBlocks ).toMatchObject( [
+				{
+					name: 'core/paragraph',
+					attributes: { content: '3rd' },
+				},
+				{
+					name: 'core/paragraph',
+					attributes: { content: '1st' },
+				},
+				{
+					name: 'core/paragraph',
+					attributes: { content: '2nd' },
+				},
+			] );
+		} );
+	} );
+
+	test.describe( 'test shortcuts handling through portals in the same tree', () => {
+		test( 'should propagate properly and duplicate selected blocks', async ( {
+			editor,
+			page,
+			pageUtils,
+		} ) => {
+			await addTestParagraphBlocks( { editor, page } );
+			// Multiselect via keyboard.
+			await pageUtils.pressKeys( 'primary+a', { times: 2 } );
+
+			await editor.clickBlockToolbarButton( 'Options' );
+			await expect(
+				page
+					.getByRole( 'menu', { name: 'Options' } )
+					.getByRole( 'menuitem', { name: 'Duplicate' } )
+			).toBeVisible();
+
+			// Duplicate via keyboard.
+			await pageUtils.pressKeys( 'primaryShift+d' );
+
+			await expect.poll( editor.getBlocks ).toMatchObject( [
```

Comment:
I'm not sure if it's an issue with the `editor.getBlocks` method or a bug in the editor, but after duplication, the newly inserted blocks `content` attribute is `undefined`.

I've omitted attributes in this test since it's only testing the event propagation via portals. I think it's worth to look into this separately.

Reply:
I can't reproduce this. The `content` attribute is correctly showing for me 🤔 .

Reply:
Try removing the `// Insert another block manually.` part of the example test.

Reply:
I can reproduce this after re-pulling the latest trunk.

It seems like a side-effect of https://github.com/WordPress/gutenberg/pull/43204. After duplication, they shared the same `RichTextData` content and triggered the loop-detection optimization of `toMatchObject`, and bailed out comparing it. I don't know why inserting a new block fixes it though.

Reply:
I guess we can add a check in `editor.getBlocks()` to automatically convert `RichTextData` to plain `string` for testing purposes.

Reply:
cc @ellatrix